### 导入库

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### 数据读取

In [2]:
file_path = "Sample-Superstore.csv"
df = pd.read_csv(file_path,encoding="gbk")

### 数据基本信息查看

In [3]:
df.head()

,csvbase_row_id,?Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,1,CA-2023-152156,2023/11/8,2023/11/11,Second Class,CG-12520,Claire Gute,Consumer,United States,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,2,CA-2023-152156,2023/11/8,2023/11/11,Second Class,CG-12520,Claire Gute,Consumer,United States,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,3,CA-2023-138688,2023/6/12,2023/6/16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,4,US-2022-108966,2022/10/11,2022/10/18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,5,US-2022-108966,2022/10/11,2022/10/18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [4]:
df.shape

(9994, 22)

原始销售数据包含9994条交易记录，共22个字段。

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   csvbase_row_id  9994 non-null   int64  
 1   ?Row ID         9994 non-null   int64  
 2   Order ID        9994 non-null   object 
 3   Order Date      9994 non-null   object 
 4   Ship Date       9994 non-null   object 
 5   Ship Mode       9994 non-null   object 
 6   Customer ID     9994 non-null   object 
 7   Customer Name   9994 non-null   object 
 8   Segment         9994 non-null   object 
 9   Country         9994 non-null   object 
 10  City            9994 non-null   object 
 11  State           9994 non-null   object 
 12  Postal Code     9994 non-null   int64  
 13  Region          9994 non-null   object 
 14  Product ID      9994 non-null   object 
 15  Category        9994 non-null   object 
 16  Sub-Category    9994 non-null   object 
 17  Product Name    9994 non-null   o

### 数据清洗与特征构造

In [6]:
df['Postal Code'] = df['Postal Code'].astype(str)

In [7]:
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

In [8]:
df.drop(columns=['csvbase_row_id'],inplace=True)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   ?Row ID        9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   object        
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [10]:
df.columns

Index(['?Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

In [11]:
repr(df.columns[0])

"'?Row ID'"

In [12]:
#清洗字段名称
df.columns = df.columns.str.replace('?', '', regex=False)

# 验证
print("清理后的列名:")
print(df.columns.tolist())
print(f"\n第一列: '{df.columns[0]}'")

清理后的列名:
['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit']

第一列: 'Row ID'


In [13]:
repr(df.columns[0])

"'Row ID'"

In [14]:
df.duplicated().sum()

np.int64(0)

没有重复值

In [15]:
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

没有缺失值

In [16]:
df.describe()

,Row ID,Order Date,Ship Date,Sales,Quantity,Discount,Profit
count,9994.000000,9994,9994,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.503502,2023-04-30 17:41:20.448268800,2023-05-04 16:42:15.921552896,229.858001,3.789574,0.156203,28.656897
min,1.000000,2021-01-03 00:00:00,2021-01-07 00:00:00,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,2022-05-23 00:00:00,2022-05-27 00:00:00,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,2023-06-26 00:00:00,2023-06-29 00:00:00,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,2024-05-14 00:00:00,2024-05-18 00:00:00,209.940000,5.000000,0.200000,29.364000
max,9994.000000,2024-12-30 00:00:00,2025-01-05 00:00:00,22638.480000,14.000000,0.800000,8399.976000
std,2885.160014,NaN,NaN,623.245101,2.225110,0.206452,234.260108


通过对销售数据进行描述性统计分析，数据集共包含9994条订单记录，涵盖2014年至2017年的销售业务数据。  
从销售指标来看，订单销售额（Sales）的平均值为229.86美元，中位数为54.49美元，二者存在较大差异，说明销售额呈现明显的右偏分布，少量高价值订单贡献了较高的销售金额。  
利润（Profit）的平均值为28.66美元，但最小值达到-6599.98美元，表明部分订单存在亏损情况。由于负利润可能反映折扣策略、产品成本或销售策略问题，因此在后续分析中不将其作为异常数据删除，而是进一步分析亏损订单的分布特征。  
折扣（Discount）的平均水平为15.62%，最高达到80%，说明不同订单之间存在较大的价格优惠差异。较高折扣可能促进销售增长，但也可能降低企业利润，因此后续将结合利润指标进一步分析折扣对盈利能力的影响。  
商品购买数量平均为3.78件，最大购买数量为14件，整体分布较为稳定，不存在明显异常情况。

In [17]:
#利润率
df['Profit Margin'] = (df['Profit'] / df['Sales'])

In [18]:
#订单年份
df['Order Year'] = (df['Order Date'].dt.year)

In [19]:
#订单月份
df['Order Month'] = (df['Order Date'].dt.month)

In [20]:
#季度
df['Order Quarter'] = (df['Order Date'].dt.quarter)

In [21]:
#订单处理时间
df['Order Processing Days'] = ((df['Ship Date'] - df['Order Date']).dt.days)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Row ID                 9994 non-null   int64         
 1   Order ID               9994 non-null   object        
 2   Order Date             9994 non-null   datetime64[ns]
 3   Ship Date              9994 non-null   datetime64[ns]
 4   Ship Mode              9994 non-null   object        
 5   Customer ID            9994 non-null   object        
 6   Customer Name          9994 non-null   object        
 7   Segment                9994 non-null   object        
 8   Country                9994 non-null   object        
 9   City                   9994 non-null   object        
 10  State                  9994 non-null   object        
 11  Postal Code            9994 non-null   object        
 12  Region                 9994 non-null   object        
 13  Pro

In [23]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Product Name,Sales,Quantity,Discount,Profit,Profit Margin,Order Year,Order Month,Order Quarter,Order Processing Days
0,1,CA-2023-152156,2023-11-08,2023-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,0.1600,2023,11,4,3
1,2,CA-2023-152156,2023-11-08,2023-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,0.3000,2023,11,4,3
2,3,CA-2023-138688,2023-06-12,2023-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,0.4700,2023,6,2,4
3,4,US-2022-108966,2022-10-11,2022-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,-0.4000,2022,10,4,7
4,5,US-2022-108966,2022-10-11,2022-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,0.1125,2022,10,4,7


### Date Warehouse Modeling

In [24]:
#新增客户维度表
dim_customer = df[
    [
        'Customer ID',
        'Customer Name',
        'Segment',
        'Country',
        'State',
        'City',
        'Postal Code',
        'Region'
    ]
].drop_duplicates(
    subset=['Customer ID']
)

In [25]:
dim_customer.head()

,Customer ID,Customer Name,Segment,Country,State,City,Postal Code,Region
0,CG-12520,Claire Gute,Consumer,United States,Kentucky,Henderson,42420,South
2,DV-13045,Darrin Van Huff,Corporate,United States,California,Los Angeles,90036,West
3,SO-20335,Sean O'Donnell,Consumer,United States,Florida,Fort Lauderdale,33311,South
5,BH-11710,Brosina Hoffman,Consumer,United States,California,Los Angeles,90032,West
12,AA-10480,Andrew Allen,Consumer,United States,North Carolina,Concord,28027,South


In [26]:
dim_customer.shape

(793, 8)

In [27]:
#创建商品维度表
dim_product = df[
    ['Product ID',
     'Product Name',
     'Category',
     'Sub-Category'
    ]
].drop_duplicates(
    subset = ['Product ID']
)

In [28]:
dim_product.head()

,Product ID,Product Name,Category,Sub-Category
0,FUR-BO-10001798,Bush Somerset Collection Bookcase,Furniture,Bookcases
1,FUR-CH-10000454,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",Furniture,Chairs
2,OFF-LA-10000240,Self-Adhesive Address Labels for Typewriters b...,Office Supplies,Labels
3,FUR-TA-10000577,Bretford CR4500 Series Slim Rectangular Table,Furniture,Tables
4,OFF-ST-10000760,Eldon Fold 'N Roll Cart System,Office Supplies,Storage


In [29]:
dim_product.shape

(1861, 4)

In [30]:
dim_product['Product ID'].nunique()

1861

In [31]:
#创建销售事实表
fact_sales = df[
[
'Row ID',
'Order ID',
'Customer ID',
'Product ID',
'Order Date',
'Sales',
'Quantity',
'Discount',
'Profit',
'Profit Margin',
'Order Processing Days'
]
].copy()

In [32]:
# 修改字段名称，使其符合数据仓库命名规范

fact_sales.rename(
    columns={
        'Row ID': 'Sales ID'
    },
    inplace=True
)

In [33]:
fact_sales.head()

,Sales ID,Order ID,Customer ID,Product ID,Order Date,Sales,Quantity,Discount,Profit,Profit Margin,Order Processing Days
0,1,CA-2023-152156,CG-12520,FUR-BO-10001798,2023-11-08,261.9600,2,0.00,41.9136,0.1600,3
1,2,CA-2023-152156,CG-12520,FUR-CH-10000454,2023-11-08,731.9400,3,0.00,219.5820,0.3000,3
2,3,CA-2023-138688,DV-13045,OFF-LA-10000240,2023-06-12,14.6200,2,0.00,6.8714,0.4700,4
3,4,US-2022-108966,SO-20335,FUR-TA-10000577,2022-10-11,957.5775,5,0.45,-383.0310,-0.4000,7
4,5,US-2022-108966,SO-20335,OFF-ST-10000760,2022-10-11,22.3680,2,0.20,2.5164,0.1125,7


In [34]:
fact_sales.shape

(9994, 11)

In [35]:
fact_sales.duplicated().sum()

np.int64(0)

In [36]:
#创建日期维度表
date_range = pd.date_range(
    start=df['Order Date'].min(),
    end=df['Order Date'].max()
)


dim_date = pd.DataFrame(
    {
        'Date': date_range
    }
)

In [37]:
dim_date.head()

,Date
0,2021-01-03
1,2021-01-04
2,2021-01-05
3,2021-01-06
4,2021-01-07


In [38]:
# 年份

dim_date['Year'] = (
    dim_date['Date'].dt.year
)


# 月份

dim_date['Month'] = (
    dim_date['Date'].dt.month
)


# 月份名称

dim_date['Month Name'] = (
    dim_date['Date'].dt.month_name()
)


# 季度

dim_date['Quarter'] = (
    dim_date['Date'].dt.quarter
)

In [39]:
dim_date['Date ID'] = (
    dim_date['Date']
    .dt.strftime('%Y%m%d')
    .astype(int)
)

In [40]:
dim_date.head()

,Date,Year,Month,Month Name,Quarter,Date ID
0,2021-01-03,2021,1,January,1,20210103
1,2021-01-04,2021,1,January,1,20210104
2,2021-01-05,2021,1,January,1,20210105
3,2021-01-06,2021,1,January,1,20210106
4,2021-01-07,2021,1,January,1,20210107


In [41]:
dim_date.shape

(1458, 6)

In [42]:
dim_date.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458 entries, 0 to 1457
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Date        1458 non-null   datetime64[ns]
 1   Year        1458 non-null   int32         
 2   Month       1458 non-null   int32         
 3   Month Name  1458 non-null   object        
 4   Quarter     1458 non-null   int32         
 5   Date ID     1458 non-null   int64         
dtypes: datetime64[ns](1), int32(3), int64(1), object(1)
memory usage: 51.4+ KB


### Data Export

In [43]:
print("客户表:", dim_customer.shape)
print("商品表:", dim_product.shape)
print("日期表:", dim_date.shape)
print("销售事实表:", fact_sales.shape)

客户表: (793, 8)
商品表: (1861, 4)
日期表: (1458, 6)
销售事实表: (9994, 11)


将各表保存为csv

In [44]:
dim_customer.to_csv(
    "dim_customer.csv",
    index=False,
    encoding="utf-8-sig"
)

In [45]:
dim_product.to_csv(
    "dim_product.csv",
    index=False,
    encoding="utf-8-sig"
)

In [46]:
dim_date.to_csv(
    "dim_date.csv",
    index=False,
    encoding="utf-8-sig"
)

In [47]:
fact_sales.to_csv(
    "fact_sales.csv",
    index=False,
    encoding="utf-8-sig"
)

In [48]:
import os

os.getcwd()

'E:\\python workspace\\企业销售数据仓库与经营分析BI系统'

### Data Table Join with SQL

In [49]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://root:yang060828@localhost:3306/superstore_bi"
)

In [50]:
dim_product.columns

Index(['Product ID', 'Product Name', 'Category', 'Sub-Category'], dtype='object')

In [51]:
dim_product.rename(
    columns={
        "Product ID":"product_id",
        "Product Name":"product_name",
        "Category":"category",
        "Sub-Category":"sub_category"
    },
    inplace=True
)

In [52]:
dim_product.to_sql(
    name="dim_product",
    con=engine,
    if_exists="replace",
    index=False
)

1861

In [53]:
dim_customer.rename(
    columns={
        "Customer ID": "customer_id",
        "Customer Name": "customer_name",
        "Segment": "segment",
        "Country": "country",
        "State": "state",
        "City": "city",
        "Postal Code": "postal_code",
        "Region": "region"
    },
    inplace=True
)

In [54]:
dim_customer.to_sql(
    "dim_customer",
    engine,
    if_exists="replace",
    index=False
)

793

In [55]:
dim_date.rename(
    columns={
        "Date": "date",
        "Year": "year",
        "Month": "month",
        "Month Name": "month_name",
        "Quarter": "quarter",
        "Date ID": "date_id"
    },
    inplace=True
)

In [56]:
dim_date.to_sql(
    name="dim_date",
    con=engine,
    if_exists="replace",
    index=False
)

1458

In [57]:
fact_sales.rename(
    columns={
        "Row ID": "sales_id",
        "Order ID": "order_id",
        "Customer ID": "customer_id",
        "Product ID": "product_id",
        "Order Date": "order_date",
        "Sales": "sales",
        "Quantity": "quantity",
        "Discount": "discount",
        "Profit": "profit",
        "Profit Margin": "profit_margin",
        "Order Processing Days": "order_processing_days"
    },
    inplace=True
)

In [58]:
fact_sales.to_sql(
    name="fact_sales",
    con=engine,
    if_exists="replace",
    index=False
)

9994